common/errors.go

```diff
@@ -5,16 +5,54 @@ import (
 	"runtime"
 )
 
+//----------------------------------------
+// Convenience methods
+
+// ErrorWrap will just call .TraceAt(), or create a new *cmnError.
+func ErrorWrap(cause interface{}, format string, a ...interface{}) Error {
+	msg := Fmt(format, a...)
+	if causeCmnError, ok := cause.(*cmnError); ok {
+		return causeCmnError.TraceAt(1, msg)
+	} else {
+		// NOTE: cause may be nil.
+		// NOTE: do not use causeCmnError here, not the same as nil.
+		return newError(msg, cause, cause).Stacktrace()
+	}
+}
+
 //----------------------------------------
 // Error & cmnError
 
+/*
+Usage:
+
+```go
+	// Error construction
+	var someType = errors.New("Some err type")
+	var err1 error = NewErrorWithType(someType, "my message")
+	...
+	// Wrapping
+	var err2 error  = ErrorWrap(err1, "another message")
+	if (err1 != err2) { panic("should be the same")
+	...
+	// Error handling
+	switch err2.Type() {
+		case someType: ...
+	    default: ...
+	}
+```
+
+*/
 type Error interface {
 	Error() string
+	Message() string
+	Stacktrace() Error
 	Trace(format string, a ...interface{}) Error
-	TraceCause(cause error, format string, a ...interface{}) Error
-	Cause() error
-	Type() interface{}
+	TraceAt(n int, format string, a ...interface{}) Error
+	Cause() interface{}
 	WithType(t interface{}) Error
+	Type() interface{}
```

Comment:
Is there a better name?  Switcher?  
`switch err.Type().(type)` is redundant...
`switch err.Switcher().(type)` is also redundant...

Reply:
I'd vote for the first option (maybe even shorten it to `err.T().(type)`. go is redundant language (`obj.(Object).SomeMethod`).

Reply:
I like "T" a lot.  It differentiates it from the "type" keyword.